test/SetPath.cpp

```diff
@@ -0,0 +1,408 @@
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NullValueAppliedWithHost) {
+	UriUriA uri = parseWellFormedUri("//host/path");
+
+	EXPECT_EQ(uriSetPathA(&uri, NULL, NULL), URI_SUCCESS);
+
+	assertUriEqual(&uri, "//host");
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NullValueAppliedWithoutHost) {
+	UriUriA uri = parseWellFormedUri("scheme:/path");
+
+	EXPECT_EQ(uriSetPathA(&uri, NULL, NULL), URI_SUCCESS);
+
+	assertUriEqual(&uri, "scheme:");
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NonNullValueAppliedSingleSlashWithHost) {
+	UriUriA uri = parseWellFormedUri("//host/path");
+	const char * const first = "/";
+	const char * const afterLast = first + strlen(first);
+
+	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);
+
+	assertUriEqual(&uri, "//host/");
+	EXPECT_EQ(uri.absolutePath, URI_FALSE);  // always false for URIs with host
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NonNullValueAppliedSingleSlashWithoutHost) {
+	UriUriA uri = parseWellFormedUri("scheme:path");
+	const char * const first = "/";
+	const char * const afterLast = first + strlen(first);
+
+	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);
+
+	assertUriEqual(&uri, "scheme:/");
+	EXPECT_EQ(uri.absolutePath, URI_TRUE);
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NonNullValueAppliedTwoSlashesWithHost) {
+	UriUriA uri = parseWellFormedUri("//host/path");
+	const char * const first = "//";
+	const char * const afterLast = first + strlen(first);
+
+	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);
+
+	assertUriEqual(&uri, "//host//");
+	EXPECT_EQ(uri.absolutePath, URI_FALSE);  // always false for URIs with host
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NonNullValueAppliedTwoSlashesWithoutHostDotInserted) {
+	UriUriA uri = parseWellFormedUri("scheme:path");
+	const char * const first = "//";
+	const char * const afterLast = first + strlen(first);
+
+	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);
+
+	assertUriEqual(&uri, "scheme:/.//");  // i.e. not scheme://
+	EXPECT_EQ(uri.absolutePath, URI_TRUE);
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NonNullValueAppliedThreeSlashesWithHost) {
+	UriUriA uri = parseWellFormedUri("//host/path");
+	const char * const first = "///";
+	const char * const afterLast = first + strlen(first);
+
+	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);
+
+	assertUriEqual(&uri, "//host///");
+	EXPECT_EQ(uri.absolutePath, URI_FALSE);  // always false for URIs with host
+
+	uriFreeUriMembersA(&uri);
+}
+
+TEST(SetPath, NonNullValueAppliedThreeSlashesWithoutHostDotInserted) {
+	UriUriA uri = parseWellFormedUri("scheme:path");
+	const char * const first = "///";
+	const char * const afterLast = first + strlen(first);
+
+	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);
+
+	assertUriEqual(&uri, "scheme:/.///");  // i.e. not scheme:///
+	EXPECT_EQ(uri.absolutePath, URI_TRUE);
+
+	uriFreeUriMembersA(&uri);
+}
```

Comment:
Should we have an explicit test case verifying the behavior for `file:` URIs? This seems to be a real world use case that folks expect to work and from what I see we don't have any path test with an empty host yet.

```
	UriUriA uri = parseWellFormedUri("file:///tmp/example.txt");
	const char * const first = "/root/example.txt";
	const char * const afterLast = first + strlen(first);

	EXPECT_EQ(uriSetPathA(&uri, first, afterLast), URI_SUCCESS);

	assertUriEqual(&uri, "file:///root/example.txt");
	EXPECT_EQ(uri.absolutePath, URI_FALSE);

	uriFreeUriMembersA(&uri);
```

Reply:
@TimWolla the `file` scheme is by no means special to RFC 3986, so it did not get any dedicated test cases. I have still added a testcase `SetPath.NonNullValueAppliedNonEmptyWithEmptyHost` for you now that "happens" to use the `file` scheme so that you can be at ease about both the empty host and `file` URIs working as expected. Push upcoming…

Reply:
> the `file` scheme is by no means special to RFC 3986

Yes, I understand. This was perhaps badly phrased from my side, the `file` scheme was just a convenient way to showcase this particular case of three consecutive slashes / empty hostnames. Thanks!

Reply:
@TimWolla I'll mark this as resolved already, please unresolve or comment again if the just-pushed test case has any issues.